flow/connectors/clickhouse/clickhouse.go

```diff
@@ -273,3 +273,136 @@ func (c *ClickhouseConnector) execWithLogging(ctx context.Context, query string)
 	c.logger.Info("[clickhouse] executing DDL statement", slog.String("query", query))
 	return c.database.Exec(ctx, query)
 }
+
+func (c *ClickhouseConnector) checkTablesEmptyAndEngine(ctx context.Context, tables []string) error {
+	queryInput := make([]interface{}, 0, len(tables)+1)
+	queryInput = append(queryInput, c.config.Database)
+	for _, table := range tables {
+		queryInput = append(queryInput, table)
+	}
+	rows, err := c.database.Query(ctx,
+		fmt.Sprintf("SELECT name,engine,total_rows FROM system.tables WHERE database=? AND table IN (%s)",
+			strings.Join(slices.Repeat([]string{"?"}, len(tables)), ",")), queryInput...)
+	if err != nil {
+		return fmt.Errorf("failed to get information for destination tables: %w", err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var tableName, engine string
+		var totalRows uint64
+		err = rows.Scan(&tableName, &engine, &totalRows)
+		if err != nil {
+			return fmt.Errorf("failed to scan information for tables: %w", err)
+		}
+		if totalRows != 0 {
+			return fmt.Errorf("table %s exists and is not empty", tableName)
+		}
+		if !slices.Contains(acceptableTableEngines, engine) {
+			return fmt.Errorf("table %s exists and is not using ReplacingMergeTree/MergeTree engine", tableName)
+		}
+	}
+	if rows.Err() != nil {
+		return fmt.Errorf("failed to read rows: %w", rows.Err())
+	}
+	return nil
+}
+
+func (c *ClickhouseConnector) getTableColumnsMapping(ctx context.Context,
+	tables []string,
+) (map[string][]*protos.FieldDescription, error) {
+	tableColumnsMapping := make(map[string][]*protos.FieldDescription, len(tables))
+	queryInput := make([]interface{}, 0, len(tables)+1)
+	queryInput = append(queryInput, c.config.Database)
+	for _, table := range tables {
+		queryInput = append(queryInput, table)
+	}
+	rows, err := c.database.Query(ctx,
+		fmt.Sprintf("SELECT name,type,table FROM system.columns WHERE database=? AND table IN (%s)",
+			strings.Join(slices.Repeat([]string{"?"}, len(tables)), ",")), queryInput...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get columns for destination tables: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var tableName string
+		var fieldDescription protos.FieldDescription
+		err = rows.Scan(&fieldDescription.Name, &fieldDescription.Type, &tableName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to scan columns for tables: %w", err)
+		}
+		tableColumnsMapping[tableName] = append(tableColumnsMapping[tableName], &fieldDescription)
+	}
+	if rows.Err() != nil {
+		return nil, fmt.Errorf("failed to read rows: %w", rows.Err())
+	}
+	return tableColumnsMapping, nil
+}
+
+func (c *ClickhouseConnector) processTableComparison(dstTableName string, srcSchema *protos.TableSchema,
+	dstSchema []*protos.FieldDescription, peerDBColumns []string,
+) error {
+	for _, srcField := range srcSchema.Columns {
+		found := false
+		for _, dstField := range dstSchema {
+			// not doing type checks for now
+			if srcField.Name == dstField.Name {
```

Comment:
#2010 makes it possible to override destination column name

Reply:
will hold off if it lands soon, else follow up